File: leetcode/leetcode-32.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        len1 = len(s)
        if len1 == 0:
            return 0
        dp = [0] *len1
        for i in range(1, len1):
            if s[i] == ')':
                if s[i-1] == '(':
                    if i == 1:
                        dp[i] = 2
                    else:
                        dp[i] = dp[i-2] + 2
                elif i-dp[i-1] > 0 and s[i-dp[i-1]-1] == '(':
                    if i - dp[i-1] > 1:
                        dp[i] = dp[i-1] + dp[i-dp[i-1]-2] + 2
                    else:
                        dp[i] = dp[i-1] + 2
        ans = max(dp)
        return ans

if __name__ == "__main__":
    so = Solution()
    s = ')()())'
    print(so.longestValidParentheses(s))

File: leetcode/leetcode-32.cpp
/*
 *@File Name          :leetcode-32.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/7/11 9:43
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "queue"
#include "string"
#include "vector"
using namespace std;

class Solution {
 public:
  int longestValidParentheses(string s) {
    int         n = s.size();
    vector<int> dp(n, 0);
    int         ans = 0;

    priority_queue<int> que;
    for (int i = 1; i < n; i++) {
      if (s[i] == ')' && i - 1 - dp[i - 1] >= 0 && s[i - 1 - dp[i - 1]] == '(')
        dp[i] += dp[i - 1] + 2;
      if (i - 1 - dp[i - 1] - 1 >= 0 && dp[i])
        dp[i] += dp[i - 1 - dp[i - 1] - 1];
      ans = max(ans, dp[i]);
    }
    return ans;
  }
};